akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala

```diff
@@ -664,14 +664,25 @@ private[http] object HttpServerBluePrint {
           })
 
           setHandler(fromNet, new InHandler {
-            override def onPush(): Unit = sourceOut.push(grab(fromNet).bytes)
+            override def onPush(): Unit = {
+              if (sourceOut.isAvailable) {
+                sourceOut.push(grab(fromNet).bytes)
+              }
+            }
             override def onUpstreamFinish(): Unit = sourceOut.complete()
             override def onUpstreamFailure(ex: Throwable): Unit = sourceOut.fail(ex)
           })
           sourceOut.setHandler(new OutHandler {
             override def onPull(): Unit = {
-              if (!hasBeenPulled(fromNet)) pull(fromNet)
+              // This check only needed on the first pull due to potential element
+              // pushed in response to pull by previous source
```

Comment:
Yep, exactly that's the reason. Good catch. Execution of subsource handlers is less deterministic than things running in a single interpreter because the exact interleaving of event processing depends on when exactly the event horizon of the participating interpreters are depleted.